scipy/stats/_binned_statistic.py

```diff
@@ -381,60 +511,65 @@ def binned_statistic_dd(sample, values, statistic='mean',
 
     nbin = np.asarray(nbin)
 
-    # Compute the bin number each sample falls into.
-    Ncount = {}
-    for i in np.arange(D):
-        Ncount[i] = np.digitize(sample[:, i], edges[i])
+    # Compute the bin number each sample falls into, in each dimension
+    sampBin = {}
+    for i in xrange(Ndim):
+        sampBin[i] = np.digitize(sample[:, i], edges[i])
 
-    # Using digitize, values that fall on an edge are put in the right bin.
+    # Using `digitize`, values that fall on an edge are put in the right bin.
     # For the rightmost bin, we want values equal to the right
     # edge to be counted in the last bin, and not as an outlier.
-    for i in np.arange(D):
-        # Rounding precision
+    for i in xrange(Ndim):
+        # Find the rounding precision
         decimal = int(-np.log10(dedges[i].min())) + 6
         # Find which points are on the rightmost edge.
-        on_edge = np.where(np.around(sample[:, i], decimal)
-                           == np.around(edges[i][-1], decimal))[0]
+        on_edge = np.where(np.around(sample[:, i], decimal) ==
+                           np.around(edges[i][-1], decimal))[0]
         # Shift these points one bin to the left.
-        Ncount[i][on_edge] -= 1
+        sampBin[i][on_edge] -= 1
 
     # Compute the sample indices in the flattened statistic matrix.
     ni = nbin.argsort()
-    xy = np.zeros(N, int)
-    for i in np.arange(0, D - 1):
-        xy += Ncount[ni[i]] * nbin[ni[i + 1:]].prod()
-    xy += Ncount[ni[-1]]
+    # `binnumbers` is which bin (in linearized `Ndim` space) each sample goes
+    binnumbers = np.zeros(Dlen, int)
+    for i in xrange(0, Ndim - 1):
+        binnumbers += sampBin[ni[i]] * nbin[ni[i + 1:]].prod()
+    binnumbers += sampBin[ni[-1]]
 
-    result = np.empty(nbin.prod(), float)
+    result = np.empty([Vdim, nbin.prod()], float)
 
     if statistic == 'mean':
         result.fill(np.nan)
-        flatcount = np.bincount(xy, None)
-        flatsum = np.bincount(xy, values)
+        flatcount = np.bincount(binnumbers, None)
         a = flatcount.nonzero()
-        result[a] = flatsum[a] / flatcount[a]
+        for vv in xrange(Vdim):
+            flatsum = np.bincount(binnumbers, values[vv])
+            result[vv, a] = flatsum[a] / flatcount[a]
```

Comment:
Why did you change the broadcasting approach to a ``for`` loop here? I'd assume the old version would be faster.